### DISCUSSION TITLE
Upgrading to 1.19 (or above)

### DISCUSSION THREAD

**Discussion by mawo66** [kimai/kimai#3414] [General]
**Describe the bug**
The update from 1.18 to 1.19 fails during `bin/console kimai:update`.

**To Reproduce**
1. git fetch --tags
2. git checkout 1.19
3. composer install --no-dev --optimize-autoloader -n
4. bin/console kimai:update

On step 4, the update fails:
```
Kimai updates running ...
=========================

[notice] Migrating up to DoctrineMigrations\Version20220315224645
[error] Migration DoctrineMigrations\Version20220315224645 failed during Execution. Error: "The column 'billable' on table 'kimai2_activities' already exists."

 [ERROR] The column 'billable' on table 'kimai2_activities' already exists.
```

**Additional context**
 - Kimai version: 1.18.2 
 - PHP version: 7.4
 - Debian 11

**Comment by kevinpapst** [comment]
I converted your issue because there is no bug and you might need some help with the update anyways... 

If you don't use the column, delete it and re-run the update.

**Comment by mawo66** [comment]
OK, thanks for the advice. I did delete the column and re-ran the update.
But now another error shows up:
```
Kimai updates running ...
=========================

[notice] Migrating up to DoctrineMigrations\Version20220315224645
[error] Migration DoctrineMigrations\Version20220315224645 failed during Execution. Error: "An exception occurred while executing 'ALTER TABLE kimai2_customers ADD billable TINYINT(1) DEFAULT '1' NOT NULL':

SQLSTATE[42000]: Syntax error or access violation: 1118 Row size too large. The maximum row size for the used table type, not counting BLOBs, is 8126. This includes storage overhead, check the manual. You have to change some columns to TEXT or BLOBs"

 [ERROR] An exception occurred while executing 'ALTER TABLE kimai2_customers ADD billable TINYINT(1)
         DEFAULT '1' NOT NULL':

         SQLSTATE[42000]: Syntax error or access violation: 1118 Row size too large. The maximum row
         size for the used table type, not counting BLOBs, is 8126. This includes storage overhead,
         check the manual. You have to change some columns to TEXT or BLOBs
```
I did never expirience any problems with the update until version 1.18.2 and I'm using kimai for almost 3 years now. BTW: Thanks for this great tool!

**Reply by kevinpapst** [reply]
I don't know what you are doing, but you might want to start from scratch.

**Reply by kevinpapst** [reply]
Read here about this issue: https://mariadb.com/kb/en/troubleshooting-row-size-too-large-errors-with-innodb/

As this does not happen on any other system, I need you to debug the issue on your end.

**Reply by mawo66** [reply]
Changing the row format of the table did the trick:
```
ALTER TABLE kimai2_customers ROW_FORMAT=DYNAMIC;
```

After that, the update finishes sucessfully. Thanks for your support!

**Reply by kevinpapst** [reply]
Can you share some details about your system? Why that happened in the first place?
Runtime configuration? Please post `DESCRIBE kimai2_customers;`
There is some problem that you now have hidden, but I do not understand why that happens.

**Comment by mawo66** [comment]
I installed kimai a few years ago on a virtual machine with Debian 9 by cloning the repository and using composer. The database I created according to the installation description. I updated kimai several times without issues. The database itself was not manipulated manually. A few month ago, I upgraded Debian to version 11. Kimai kept running, so I kept updating it until 1.18.2 was reached. When I tried the update to 1.19 the problem occured.

```
MariaDB [kimai2]> DESCRIBE kimai2_customers;
+-------------+--------------+------+-----+---------+----------------+
| Field       | Type         | Null | Key | Default | Extra          |
+-------------+--------------+------+-----+---------+----------------+
| id          | int(11)      | NO   | PRI | NULL    | auto_increment |
| name        | varchar(150) | NO   |     | NULL    |                |
| number      | varchar(50)  | YES  |     | NULL    |                |
| comment     | text         | YES  |     | NULL    |                |
| visible     | tinyint(1)   | NO   | MUL | NULL    |                |
| company     | varchar(255) | YES  |     | NULL    |                |
| contact     | varchar(255) | YES  |     | NULL    |                |
| address     | text         | YES  |     | NULL    |                |
| country     | varchar(2)   | NO   |     | NULL    |                |
| currency    | varchar(3)   | NO   |     | NULL    |                |
| phone       | varchar(255) | YES  |     | NULL    |                |
| fax         | varchar(255) | YES  |     | NULL    |                |
| mobile      | varchar(255) | YES  |     | NULL    |                |
| email       | varchar(255) | YES  |     | NULL    |                |
| homepage    | varchar(255) | YES  |     | NULL    |                |
| timezone    | varchar(64)  | NO   |     | NULL    |                |
| color       | varchar(7)   | YES  |     | NULL    |                |
| time_budget | int(11)      | NO   |     | 0       |                |
| budget      | double       | NO   |     | 0       |                |
| vat_id      | varchar(50)  | YES  |     | NULL    |                |
| budget_type | varchar(10)  | YES  |     | NULL    |                |
| billable    | tinyint(1)   | NO   |     | 1       |                |
+-------------+--------------+------+-----+---------+----------------+
````

**Reply by kevinpapst** [reply]
Thank @mawo66 for the feedback!

I just checked the docs https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-utf8mb4.html as yes, a UTF-8 character might fill up 4 byte. Calculating that from the columns `(10+50+11+7+64+255+255+255+255+255+255+255+150+50)*4` is indeed > 8500. Extremely unlikely that someone fills all available to the end with UTF8 characters, but maybe the strict mode calculates the possible max size and not the real used one.

Seems I have to do some reasonable column downsizing for the strict mode. These large columns are still a sin from day-one 😁 so they are pretty old and obviously 255 character for a phone or fax number are not necessary. 

Ok, what do you think about these new sizes?

| Field       | old length         | new length |
|----------|--------------|----------------------|
| company     | 255 | 100 |
| contact     | 255 | 50 |
| phone       | 255 | 30 |
| fax         | 255 | 30 |
| mobile      | 255 | 30 |
| email       | 255 | 255 |
| homepage    | 255 | 100 |

Phone numbers are 15 character max (see [wikipedia](https://en.wikipedia.org/wiki/E.164)), but people might write them with separators/spaces and such, so I'd propose 30 character to make sure it fits. 

Not sure about the email. Technically it is limited to 254 character, but I have never seen an email that long.

**Reply by mawo66** [reply]
I think the new sizes of the fields are absolutely reasonable.
But I still don't understand, why the problem apparently only showed up on my installation. Do you have a theory about that?

**Reply by kevinpapst** [reply]
Me neither, and it is very strange that the error suddenly pops up, when a one byte field should be added.
To me that sounds as if you changed something since the last Kimai update (either a database update or database configuration change).
